### ISSUE TITLE
Update SAM 3 Integration for Local Model Inference

### ISSUE THREAD

**Issue by sean-roth** [sean-roth/promptmask#2]
## Overview

Update the SAM 3 model integration to use Hugging Face Transformers library for local model download and inference. This enables the application to run completely locally on the user's GPU without requiring API calls.

## Context

SAM 3 was released on November 19, 2025 and is available through Hugging Face Transformers. The current implementation uses a custom model loader that needs to be replaced with the official Transformers API for:

- ✅ One-time model download (~3GB to local cache: `~/.cache/huggingface/hub/`)
- ✅ Pure local inference on GPU/CPU (no API calls, no costs)
- ✅ Automatic device detection (CUDA/MPS/CPU)
- ✅ Works offline after initial download

### 1. Update `sam3/model_loader.py`

Replace the current custom loader with Transformers-based loading:

```python
"""
SAM 3 Model Loader
Downloads and loads SAM 3 model locally using Hugging Face Transformers.
"""

import logging
import torch
from transformers import Sam3Model, Sam3Processor

logger = logging.getLogger(__name__)


class SAM3ModelLoader:
    """Handles loading and caching of SAM 3 model."""

    def __init__(self):
        """Initialize model loader."""
        self.model = None
        self.processor = None
        self.device = self._get_device()
        logger.info(f"Initialized SAM3ModelLoader with device: {self.device}")

    def _get_device(self) -> torch.device:
        """
        Auto-detect best available device.
        
        Returns:
            torch.device: Best available device (CUDA > MPS > CPU)
        """
        if torch.cuda.is_available():
            device = torch.device("cuda")
            logger.info(f"Using CUDA GPU: {torch.cuda.get_device_name(0)}")
        elif torch.backends.mps.is_available():
            device = torch.device("mps")
            logger.info("Using Apple Silicon GPU (MPS)")
        else:
            device = torch.device("cpu")
            logger.info("Using CPU (no GPU available)")
        
        return device

    def load_model(self):
        """
        Download and load SAM 3 model locally.
        
        First run downloads ~3GB to ~/.cache/huggingface/hub/
        Subsequent runs use cached model.
        
        Returns:
            tuple: (model, processor)
        """
        logger.info("Loading SAM 3 model from Hugging Face...")
        
        try:
            # Download and load model (cached after first run)
            self.model = Sam3Model.from_pretrained("facebook/sam3").to(self.device)
            self.processor = Sam3Processor.from_pretrained("facebook/sam3")
            
            logger.info("SAM 3 model loaded successfully!")
            logger.info(f"Model has {sum(p.numel() for p in self.model.parameters()) / 1e9:.2f}B parameters")
            
            return self.model, self.processor
            
        except Exception as e:
            logger.error(f"Failed to load SAM 3 model: {e}")
            logger.error("Make sure you're authenticated with Hugging Face:")
            logger.error("  Run: huggingface-cli login")
            raise

    def get_model(self):
        """Get loaded model."""
        if self.model is None:
            raise RuntimeError("Model not loaded. Call load_model() first.")
        return self.model

    def get_processor(self):
        """Get loaded processor."""
        if self.processor is None:
            raise RuntimeError("Processor not loaded. Call load_model() first.")
        return self.processor

    def get_device(self):
        """Get device being used."""
        return self.device

    def clear_cache(self):
        """Clear GPU memory cache."""
        if torch.cuda.is_available():
            torch.cuda.empty_cache()
            logger.info("Cleared CUDA cache")
```

### 2. Update `sam3/inference.py`

Update inference methods to use the official Transformers API:

```python
def segment_image(
    self,
    image: Image.Image,
    text_prompt: str,
    confidence_threshold: float = 0.7
) -> Tuple[np.ndarray, float]:
    """
    Segment an image using a text prompt - runs locally on GPU/CPU.

    Args:
        image: PIL Image to segment
        text_prompt: Natural language description of what to segment
        confidence_threshold: Minimum confidence score (0-1)

    Returns:
        Tuple of (mask as numpy array, confidence score)
        mask: Binary mask where 1 = object, 0 = background
        confidence: Confidence score for the best mask
    """
    try:
        # Prepare inputs for local inference
        inputs = self.processor(
            images=image,
            text=text_prompt,
            return_tensors="pt"
        ).to(self.device)

        # Run local inference (no API call)
        with torch.no_grad():
            outputs = self.model(**inputs)

        # Post-process results using official API
        results = self.processor.post_process_instance_segmentation(
            outputs,
            threshold=confidence_threshold,
            mask_threshold=0.5,
            target_sizes=inputs.get("original_sizes").tolist()
        )[0]

        # Extract best mask if any found
        if len(results['masks']) > 0:
            best_mask = results['masks'][0].cpu().numpy().astype(np.uint8)
            best_score = float(results['scores'][0].item())
            
            logger.debug(
                f"Segmented image with prompt '{text_prompt}' - "
                f"found {len(results['masks'])} objects, best confidence: {best_score:.2f}"
            )
            
            return best_mask, best_score
        else:
            logger.warning(
                f"No objects found for prompt '{text_prompt}' "
                f"above confidence threshold {confidence_threshold}"
            )
            # Return empty mask
            h, w = image.size[1], image.size[0]
            return np.zeros((h, w), dtype=np.uint8), 0.0

    except Exception as e:
        logger.error(f"Error during inference: {e}")
        # Return empty mask on error
        h, w = image.size[1], image.size[0]
        return np.zeros((h, w), dtype=np.uint8), 0.0
```

### 3. Update `requirements.txt`

Ensure all required packages are listed:

```
gradio==4.44.0
torch==2.3.0
torchvision==0.18.0
transformers==4.46.0
huggingface-hub==0.24.0
accelerate==0.34.0
Pillow==10.4.0
opencv-python==4.10.0
ffmpeg-python==0.2.0
numpy==1.26.4
pathlib2==2.3.7
tqdm==4.66.1
scipy==1.11.4
```

### 4. Add Setup Instructions to `INSTALLATION.md`

Add authentication section:

```markdown
## Hugging Face Authentication (Required)

SAM 3 requires authentication to download the model (gated access by Meta).

### One-Time Setup

1. Create a Hugging Face account at https://huggingface.co
2. Request access to SAM 3 at https://huggingface.co/facebook/sam3
3. Wait for approval (usually 24 hours)
4. Generate a read token at https://huggingface.co/settings/tokens
5. Authenticate on your machine:

```bash
huggingface-cli login
# Paste your token when prompted
```

### First Run

The first time you run PromptMask, it will download the SAM 3 model (~3GB) to:
- **Linux/Mac**: `~/.cache/huggingface/hub/`
- **Windows**: `C:\Users\<username>\.cache\huggingface\hub\`

This takes 5-10 minutes on a fast connection. Subsequent runs use the cached model and start instantly.

## Verifying Installation

Test the setup:

```python
from transformers import Sam3Model, Sam3Processor
import torch

# Should download model on first run
model = Sam3Model.from_pretrained("facebook/sam3")
processor = Sam3Processor.from_pretrained("facebook/sam3")
print("✅ SAM 3 loaded successfully!")
```
```

### 5. Update Main App Status Display

In `app.py`, update the model loading status to show download progress:

```python
def load_model(self):
    """Load SAM 3 model with progress display."""
    try:
        status_text = "Downloading SAM 3 model (~3GB)...\n"
        status_text += "This only happens on first run."
        gr.Info(status_text)
        
        self.model_loader.load_model()
        
        return "✅ Model loaded successfully!"
    except Exception as e:
        error_msg = f"❌ Failed to load model: {str(e)}\n\n"
        error_msg += "Make sure you're authenticated:\n"
        error_msg += "Run: huggingface-cli login"
        return error_msg
```

## Testing Checklist

After making these changes, verify:

- [ ] First run downloads model to cache (~3GB)
- [ ] Subsequent runs load instantly from cache
- [ ] Text prompts produce segmentation masks
- [ ] Model runs on available GPU (CUDA/MPS)
- [ ] Confidence threshold filtering works
- [ ] Error handling works without authentication
- [ ] Progress callbacks function correctly
- [ ] All presets (Speaker Isolation, Product Demo) work

## Benefits

✅ **No API Costs**: Runs completely locally on user's hardware
✅ **Fast Inference**: GPU-accelerated, no network latency
✅ **Offline Capable**: Works without internet after initial download
✅ **Privacy**: Video data never leaves user's machine
✅ **Scalable**: Can process unlimited videos without rate limits

## References

- SAM 3 Model: https://huggingface.co/facebook/sam3
- Official GitHub: https://github.com/facebookresearch/sam3
- Transformers Docs: https://huggingface.co/docs/transformers/main/model_doc/sam3

@claude Please implement these changes to enable local SAM 3 inference.
